workflow/controller/operator.go

```diff
@@ -513,7 +510,7 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 		woc.markWorkflowError(ctx, err)
 	}
 
-	if woc.execWf.Spec.Metrics != nil {
+	if woc.execWf.Spec.Metrics != nil && woc.wf.Status.Fulfilled() {
```

Comment:
nit: If it were previously assumed that this line until the end of the function should only be invoked after everything was officially completed, then perhaps we just add a line right above here to terminate early if we're done instead of having the condition here. What do you think? This function is so long and windy that I find myself looking for any way to improve readability and maintainability. :)